### ISSUE TITLE
Subquery changes the data type from non nullable to nullable

### ISSUE THREAD

**Issue by avinzhang** [ClickHouse/ClickHouse#70290]
**Describe the unexpected behaviour**
The original data type is not nullable, but running a subquery to aggregate the data becomes nullable

**How to reproduce**

```
create table t (id String, timestamp Datetime64) order by id;
insert into t (id, timestamp) values ('a', now());
```

The below query shows the timestamp column becomes nullable
```
select (select max(timestamp) from t) as value, toTypeName(value) from t;
   ┌───────────────────value─┬─toTypeName(value)───────┐
1. │ 2024-10-02 00:07:22.000 │ Nullable(DateTime64(3)) │
   └─────────────────────────┴─────────────────────────┘
```

* Which ClickHouse server version to use
Any version of Clickhouse

**Expected behavior**
Expect the original data type not changed

**Comment by den-crane** [comment]
v21.9.1.8000-prestable.md:* Now, scalar subquery always returns `Nullable` result if it's type can be `Nullable`. It is needed because in case of empty subquery it's result should be `Null`. Previously, it was possible to get error about incompatible types (type deduction does not execute scalar subquery, and it could use not-nullable type). Scalar subquery with empty result which can't be converted to `Nullable` (like `Array` or `Tuple`) now throws error. Fixes [#25411](https://github.com/ClickHouse/ClickHouse/issues/25411). [#26423](https://github.com/ClickHouse/ClickHouse/pull/26423) ([Nikolai Kochetov](https://github.com/KochetovNicolai)).

**Comment by Algunenano** [comment]
As Den mentions, this is the expected behaviour. You can cast it if you want to, even using assumeNotNull if you are really sure it's not null (but this is a using a shotgun so making any mistake is troublesome)
